### ISSUE TITLE
Error when  using the TensorFlow Runtime with HVX Acceleration

### ISSUE THREAD

**Issue by Geekking** [tensorflow/tensorflow#8268]
OS: Ubuntu 14.04 64bits
Android Version: 6.0.1     
NDK Version: android-ndk-r12b
I follow the description in [**tensorflow/tree/master/tensorflow/contrib/hvx**]((https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/hvx)),  after compiling the `so` files and the **hexagon_graph_execution**,  what's more, I also have pushed  additional network define file(`tensorflow_inception_v3_stripped_optimized_quantized.pb`) and the test image to my phone. 
But when I start to run the binary file in my **Snapdragon 820 android devices**, it failed all the test cases.
The stdout log is showed below,   How can I resolve this question?
 

Running main() from test_main.cc
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from GraphTransferer
[ RUN      ] GraphTransferer.RunInceptionV3OnHexagonExampleWithTfRuntime
native : hexagon_graph_execution_test.cc:294 Fuse and run inception v3 on hexagon with tf runtime
tensorflow/core/kernels/hexagon/hexagon_graph_execution_test.cc:70: Failure
Expected: (version) >= (1), actual: 0 vs 1
native : hexagon_graph_execution_test.cc:121 Read /data/local/tmp/img_299x299.bmp, size = 269156bytes
native : hexagon_graph_execution_test.cc:127 header size = 54
native : hexagon_graph_execution_test.cc:129 image size = 40
native : hexagon_graph_execution_test.cc:131 width = 299
native : hexagon_graph_execution_test.cc:133 height = -299
native : hexagon_graph_execution_test.cc:306 Ioading image finished.
native : hexagon_graph_execution_test.cc:313 Build fused graph
can't determine number of CPU cores: assuming 4
can't determine number of CPU cores: assuming 4
native : hexagon_graph_execution_test.cc:121 Read /data/local/tmp/img_299x299.bmp, size = 269156bytes
native : hexagon_graph_execution_test.cc:127 header size = 54
native : hexagon_graph_execution_test.cc:129 image size = 40
native : hexagon_graph_execution_test.cc:131 width = 299
native : hexagon_graph_execution_test.cc:133 height = -299
native : hexagon_graph_execution_test.cc:218 Ioading image finished.
native : hexagon_graph_execution_test.cc:223 Copy data to tensor.
native : hexagon_graph_execution_test.cc:246 Run graph
Init hexagon with max attributes (Controller version = 91)
Failed to disable DSP DCVS: ffffffff

Failed to append const node 65538
Failed to append const node 65538
Failed to append const node 65539
Failed to append const node 65539
.
.
.
Failed to append const node 66635
Failed to append const node 66640
Failed to append const node 66640
native : hexagon_control_wrapper.cc:235 Setup graph completed
Prepare failed! returned 0xffffffff

Execute graph!
Execution failed!
execute got err: -1

Execution failed
Failed to read data.
native : hexagon_graph_execution_test.cc:253 Output byte size = 4032
native : hexagon_graph_execution_test.cc:254 Output shape = [1,1,1,1008]
native : graph_transfer_utils.cc:43 === Dump ranking ===
native : graph_transfer_utils.cc:46 0: 1000, dumbbell, 0
native : graph_transfer_utils.cc:46 1: 999, carbonara, 0
native : graph_transfer_utils.cc:46 2: 998, stole, 0
native : graph_transfer_utils.cc:46 3: 997, rubber eraser, 0
native : graph_transfer_utils.cc:46 4: 996, coffee mug, 0
native : graph_transfer_utils.cc:46 5: 995, flagpole, 0
native : graph_transfer_utils.cc:46 6: 994, parallel bars, 0
native : graph_transfer_utils.cc:46 7: 993, cheeseburger, 0
native : graph_transfer_utils.cc:46 8: 992, bubble, 0
native : graph_transfer_utils.cc:46 9: 991, beaker, 0
Finalize hexagon
[  FAILED  ] GraphTransferer.RunInceptionV3OnHexagonExampleWithTfRuntime (5573 ms)
[----------] 1 test from GraphTransferer (5573 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (5573 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] GraphTransferer.RunInceptionV3OnHexagonExampleWithTfRuntime

 1 FAILED TEST
  YOU HAVE 3 DISABLED TESTS

**Comment by Geekking** [comment]
@tensorflower-gardener

**Comment by prb12** [comment]
@satok16 - you seem to be the owner of this contrib component.

**Comment by satok16** [comment]
> tensorflow/core/kernels/hexagon/hexagon_graph_execution_test.cc:70: Failure
> Expected: (version) >= (1), actual: 0 vs 1
This line indicates that TF can't call a binary for hexagon.  So, please answer following debugging questions.

- What kind of Android device do you have?  Basically, your Android device should have a special configuration to bypass qualcomm signature check.  This functionality can't be enabled by software but you should use some specific hardwares like a dragon board.

- What adb says?  Please attach the output from "adb logcat".

**Comment by ezhuei** [comment]
I met the same issue with 
Failed to disable DSP DCVS: ffffffff

Failed to append const node 65538
.
.
.

I tried to install the testsig library as mentioned in 
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/hvx
and it solved the issue.

**Comment by satok16** [comment]
Yeah, that's correct.  Actually I've added a check whether testsig is installed or not in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/makefile/samples/build_and_run_inception_hexagon.sh#L54.  So, if using this but still seeing the same problem, that might be a hardware restriction.

**Comment by Geekking** [comment]
@ezhuei , installing the testsig library solved this issue. 
@satok16  sorry for missing the **Troubleshooting** section in https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/hvx.

**Comment by Geekking** [comment]
Hello, @satok16 , thanks for contributing the hexagon HVX code to tensorflow.
  
I follow the tensorflow official quantization process, https://www.tensorflow.org/performance/quantization, to quantize the inception v3 model, however, it failed the test cases of hexagon graph execution, which is showed bellow. But it did pass the test cases if changed the model to https://storage.googleapis.com/download.tensorflow.org/models/tensorflow_inception_v3_stripped_optimized_quantized.pb.

[ RUN      ] GraphTransferer.RunInceptionV3OnHexagonExampleWithHexagonWrapper
native : hexagon_graph_execution_test.cc:450 Run inception v3 on hexagon with hexagon controller
native : hexagon_graph_execution_test.cc:88 Hexagon controller version is 90
can't determine number of CPU cores: assuming 4
can't determine number of CPU cores: assuming 4
tensorflow/core/kernels/hexagon/hexagon_graph_execution_test.cc:466: Failure
Value of: status.ok()
  Actual: false
Expected: true
Invalid argument: No OpKernel was registered to support Op 'DecodeJpeg' with these attrs.  Registered devices: [CPU], Registered kernels:
  <no registered kernels>

	 [[Node: DecodeJpeg = DecodeJpeg[acceptable_fraction=1, channels=3, dct_method="", fancy_upscaling=true, ratio=1, try_recover_truncated=false](DecodeJpeg/contents)]]

[  FAILED  ] GraphTransferer.RunInceptionV3OnHexagonExampleWithHexagonWrapper (331 ms)
[ RUN      ] GraphTransferer.RunInceptionV3OnHexagonExampleWithTfRuntime
native : hexagon_graph_execution_test.cc:531 Fuse and run inception v3 on hexagon with tf runtime
native : hexagon_graph_execution_test.cc:88 Hexagon controller version is 90
native : hexagon_graph_execution_test.cc:138 Read /data/local/tmp/img_299x299.bmp, size = 269156bytes
native : hexagon_graph_execution_test.cc:144 header size = 54
native : hexagon_graph_execution_test.cc:146 image size = 40
native : hexagon_graph_execution_test.cc:148 width = 299
native : hexagon_graph_execution_test.cc:150 height = -299
native : hexagon_graph_execution_test.cc:543 Ioading image finished.
native : hexagon_graph_execution_test.cc:550 Build fused graph
native : graph_transfer_utils.cc:96 Check failed: status.ok() 
Aborted 

I wonder how to quantize my model so that it will meet your hexagon code. It seems tensorflow has updated the code to quantize the model.

**Comment by satok16** [comment]
Hi Gekking,
"DecodeJpeg" is not supported yet on hexagon.  We are going to add a functionality to automatically switch implementation to cpu op-kernel implementation if the hexagon kernel is not supported.  For the time being, you should manually fuse a subgraph into a remote fused graph execute op which you want to run on hexagon.  Note that the hexagon executor only supports about 20 op kernels listed [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/hexagon/hexagon_ops_definitions.cc).  So you should avoid fusing ops which are not listed there.  See a [utility](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/hexagon/graph_transfer_utils.cc#L86) to fuse a graph.

**Comment by Geekking** [comment]
Would you mind telling me how do you quantize the test model? As in my case, I try to train a model for mnist classification dataset, I have followed the quantization tool provided by Tensorflow, the result predicted by the quantized model runned in PC is correct, however, when runned in my Mobile phone with a Snapdragon 820 devices, it got different and wrong predict score. Thanks. @satok16

**Comment by Geekking** [comment]
The model I used is visulzed ![graph_1](https://cloud.githubusercontent.com/assets/5322311/24087189/d94fcd12-0d56-11e7-90a0-ca923dbd9717.png)

![conv_1](https://cloud.githubusercontent.com/assets/5322311/24087186/d1424384-0d56-11e7-846b-bf8836192cd0.png)

**Comment by satok16** [comment]
Could you share 1. your graph file and 2. a command line which you used?

**Comment by Geekking** [comment]
I build the quantization tools on tensorflow v0.11.0 https://github.com/tensorflow/tensorflow/tree/v0.11.0. 
The graph pb file and the sh file to run the quantization process is included in 
[quantize_graph.zip](https://github.com/tensorflow/tensorflow/files/856868/quantize_graph.zip). 
Thanks.

**Comment by milinddeore** [comment]
I could able to run inception graph(tensorflow_inception_v3_stripped_optimized_quantized.pb) using the mentioned guidelines on [page](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/makefile/samples/build_and_run_inception_hexagon.sh) for Android demo TF-Classifiy application.

My target platform was OnePlus3T (which has Snapdragon 821 SOC) and i could able to achieve **inference time = 200-300ms**. I fell this is because of HVX acceleration but not quite sure and therefore i connected snapdragon's [profiler](https://developer.qualcomm.com/software/snapdragon-profiler) to check if this is indeed coming from DSP. I noticed a spike when the application starts and later very small waveforms, so there is no way to relate this to my application (at least i do not know). Is there a better way to check the performance or a quick way to know if my application is indeed using HVX. 

Another input, i checked logcat output but it doesn't show any logs coming out from 'adsp'. 

Any help is highly appreciated. Thanks.

**Comment by satok16** [comment]
Hi @milinddeore,
The first iteration may take time to determine min and max of quantized ops.  There are several ways to check whether your app is properly calling HVX functions.  The easiest way is 
```
  HexagonControlWrapper hexagon_control_wrapper;
  const int version = hexagon_control_wrapper.GetVersion();
```
If this GetVersion returns 1 or larger value, you can properly calling hvx code.  Let me know if it doesn't solve your problem.

**Comment by milinddeore** [comment]
@satok16 Thanks for info, appreciate your help. 

I have a follow up question. 
API `RunInferenceByHexagonControlWrapper()` in test section, calls  
```

// For initialisation 
Hexagon_control_wrapper.Init();
Hexagon_control_wrapper.SetUpfGraph();

and similarly 

// For Teardown
Hexagon_control_wrapper.TearDownGrpah()
Hexagon_control_wrapper.Finalize();

```
Can't we do this once at the startup and shutdown respectively? The reason i am asking, I have a phone camera which is sending images in real-time for inference and for every image i shouldn't be doing this, isn't it? Please correct me if i am wrong. Thanks.

**Comment by satok16** [comment]
You are right.  The following functions should be called only once for repeated inferences in your app.
```
Hexagon_control_wrapper.Init();
Hexagon_control_wrapper.SetUpfGraph();
Hexagon_control_wrapper.TearDownGrpah()
Hexagon_control_wrapper.Finalize();
```

**Comment by milinddeore** [comment]
@satok16 
I had written Android application for OnePlus 3T phone (This has snapdragon 821 and i have flashed my own compiled android image out of LineageOS, that means it completely rooted and /system is write accessible) but it was crashing and hence i tried following sample test for hexagon:

```
$ adb shell 'LD_LIBRARY_PATH=/data/local/tmp:$LD_LIBRARY_PATH' "/data/local/tmp/hexagon_graph_execution"
WARNING: linker: Warning: unable to normalize ""
Running main() from test_main.cc
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from GraphTransferer
[ RUN      ] GraphTransferer.RunInceptionV3OnHexagonExampleWithTfRuntime
native : hexagon_graph_execution_test.cc:446 Fuse and run inception v3 on hexagon with tf runtime
tensorflow/core/kernels/hexagon/hexagon_graph_execution_test.cc:71: Failure
Expected: (version) >= (1), actual: 0 vs 1

```
The problem here is that the version is showing '0', does that mean DSP is not accessible?

**Comment by satok16** [comment]
Unfortunately, you can't run any hexagon binary with a test signature on a production phone even if it's rooted.  You should get a development kit like https://www.intrinsyc.com/snapdragon-embedded-development-kits/snapdragon-820-development-kit/ to develop it.

**Comment by milinddeore** [comment]
Thanks @satok16!
But what if when someone try to go from development to production? Do we need to add libhexagon_nn_skel.so to the android stock and compile it as [signed-production](https://wiki.lineageos.org/signing_builds.html) build? Would that work on hexagon?

**Comment by satok16** [comment]
Hi @milinddeore,
The hexagon binary should be signed in a different way apart from the Android signature process.  So, if you want to run the hexagon binary on a phone, the hexagon binary should be signed by qualcomm or phone manufactures.

**Comment by kdxtb** [comment]
hi @satok16,

I also got some problems in using HVX accelaration.  
here are the tools and enviroments info:
Hardware: Intrinsyc’s Open‐Q™ 820 Development Kit
OS: Ubuntu 16.04 64bits
Android Version: 6.0
Hexagon SDK 3.0
NDK Version: android-ndk-r10d  as tool in Hexagon SDK
I also follow the description in tensorflow/tree/master/tensorflow/contrib/hvx, running script like this 'QUALCOMM_SDK="/home/tff/Qualcomm/Hexagon_SDK/3.0" NDK_ROOT="/home/tff/Qualcomm/Hexagon_SDK/3.0/tools/android-ndk-r10d" ./tensorflow/contrib/makefile/samples/build_and_run_inception_hexagon.sh -p'. But it failed, what can I do to get it right?

The printed log:
Running main() from test_main.cc
Note: Google Test filter = GraphTransferer.RunInceptionV3OnHexagonExampleWithTfRuntime
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from GraphTransferer
[ RUN      ] GraphTransferer.RunInceptionV3OnHexagonExampleWithTfRuntime
native : hexagon_graph_execution_test.cc:446 Fuse and run inception v3 on hexagon with tf runtime
native : hexagon_graph_execution_test.cc:72 Hexagon controller version is 90
native : hexagon_graph_execution_test.cc:122 Read /data/local/tmp/img_299x299.bmp, size = 269156bytes
native : hexagon_graph_execution_test.cc:128 header size = 54
native : hexagon_graph_execution_test.cc:130 image size = 40
native : hexagon_graph_execution_test.cc:132 width = 299
native : hexagon_graph_execution_test.cc:134 height = -299
native : hexagon_graph_execution_test.cc:458 Ioading image finished.
native : hexagon_graph_execution_test.cc:465 Build fused graph
can't determine number of CPU cores: assuming 4
can't determine number of CPU cores: assuming 4
native : op_def_util.cc:332 Op PlaceholderV2 is deprecated. It will cease to work in GraphDef version 23. Placeholder now behaves the same as PlaceholderV2..
native : hexagon_graph_execution_test.cc:122 Read /data/local/tmp/img_299x299.bmp, size = 269156bytes
native : hexagon_graph_execution_test.cc:128 header size = 54
native : hexagon_graph_execution_test.cc:130 image size = 40
native : hexagon_graph_execution_test.cc:132 width = 299
native : hexagon_graph_execution_test.cc:134 height = -299
native : hexagon_graph_execution_test.cc:262 Ioading image finished.
native : hexagon_graph_execution_test.cc:170 Ioading image finished.
native : hexagon_graph_execution_test.cc:174 Copy data to tensor.
native : hexagon_graph_execution_test.cc:284 Run graph
Init hexagon with max attributes (Controller version = 91)
native : hexagon_control_wrapper.cc:252 Setup graph completed
Prepare failed! returned 0xffffffff

Execute graph!
Execution failed!
execute got err: -1

Execution failed
Failed to read data.
native : hexagon_graph_execution_test.cc:291 Output byte size = 4032
native : hexagon_graph_execution_test.cc:292 Output shape = [1,1008]
native : graph_transfer_utils.cc:46 === Dump ranking ===
native : graph_transfer_utils.cc:49 0: 1000, dumbbell, 0
native : graph_transfer_utils.cc:49 1: 999, carbonara, 0
native : graph_transfer_utils.cc:49 2: 998, stole, 0
native : graph_transfer_utils.cc:49 3: 997, rubber eraser, 0
native : graph_transfer_utils.cc:49 4: 996, coffee mug, 0
native : graph_transfer_utils.cc:49 5: 995, flagpole, 0
native : graph_transfer_utils.cc:49 6: 994, parallel bars, 0
native : graph_transfer_utils.cc:49 7: 993, cheeseburger, 0
native : graph_transfer_utils.cc:49 8: 992, bubble, 0
native : graph_transfer_utils.cc:49 9: 991, beaker, 0
Finalize hexagon
[       OK ] GraphTransferer.RunInceptionV3OnHexagonExampleWithTfRuntime (6799 ms)
[----------] 1 test from GraphTransferer (6801 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (6803 ms total)
[  PASSED  ] 1 test.

**Comment by kdxtb** [comment]
run build_and_run_inception_hexagon.sh wih '-p' needs old version tensorflow to create hexagon_graph_execution, otherwise remove '-p',totally build all the components.

**Comment by joonblue** [comment]
Hi @satok16, @kdxtb,
I have the same problem as kdxtb has even though I didn't use -p as an option.

Prepare failed! returned 0xffffffff

NN Id = 872478096
Execute graph!
Execution failed!
execute got err: -1

NN Id = 872478096
Execution failed
NN Id = 872478096
Failed to read data.

I think that I have problem creating graph from *.pb. Do you have any suggestions?

**Comment by fly2mun** [comment]
Hi @satok16 
I try to hexagon DSP for android.
build_and_run_inception_hexagon.sh wih '-p' prebuild so files test ok.
but local build ibhexagon_nn_skel.so, this file seem to be a problem.
I thank that difference between the google server prebuild version and tensorflow source base,
When was the prebuild version created? i want to test if back.
Please let me know git commit.
Thanks, Moon.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by michaelisard** [comment]
Closing since most of this discussion seems better suited to stackoverflow. Please open a new issue if there's a specific bug or feature request.
